### PR TITLE
Fix Kafka Persistence (`WithDataVolume` and `WithDataBindMount`)

### LINKED ISSUE #4909: Kafka WithDataVolume and WithDataBindMount don't persist messages
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Calling WithDataVolume or WithDataBindMount on a Kafka resource doesn't persist data across runs of the application. We should ensure we have the right settings set up so the data gets persisted across app runs.

### Expected Behavior

When using WithDataVolume or WithDataBindMount, the messages that are produced should be perserved across app runs. So when the app spins up again, consuming those messages should be possible.

### Steps To Reproduce

Run the following test

```csharp
    [Theory]
    [InlineData(true)]
    [InlineData(false)]
    [RequiresDocker]
    public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
    {
        var cts = new CancellationTokenSource(TimeSpan.FromMinutes(3));
        var topic = "test-topic";
        string? volumeName = null;
        string? bindMountPath = null;

        try
        {
            var builder1 = CreateDistributedApplicationBuilder();
            var kafka1 = builder1.AddKafka("kafka");

            if (useVolume)
            {
                // Use a deterministic volume name to prevent them from exhausting the machines if deletion fails
                volumeName = VolumeNameGenerator.CreateVolumeName(kafka1, nameof(WithDataShouldPersistStateBetweenUsages));

                // if the volume already exists (because of a crashing previous run), try to delete it
                DockerUtils.AttemptDeleteDockerVolume(volumeName);
                kafka1.WithDataVolume(volumeName);
            }
            else
            {
                bindMountPath = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
                kafka1.WithDataBindMount(bindMountPath);
            }

            using (var app = builder1.Build())
            {
                await app.StartAsync();
                try
                {
                    var hb = Host.CreateApplicationBuilder();

                    hb.Configuration.AddInMemoryCollection(new Dictionary<string, string?>
                    {
                        [$"ConnectionStrings:{kafka1.Resource.Name}"] = await kafka1.Resource.ConnectionStringExpression.GetValueAsync(default)
                    });

        hb.AddKafkaProducer<string, string>("kafka");

                    using (var host = hb.Build())
                    {
                        await host.StartAsync();

                        var producer = host.Services.GetRequiredService<IProducer<string, string>>();
                        for (var i = 0; i < 10; i++)
                        {
                            await producer.ProduceAsync(topic, new Message<string, string> { Key = "test-key", Value = $"test-value{i}" });
                        }
                    }
                }
                finally
                {
                    // Stops the container, or the Volume/mount would still be in use
                    await app.StopAsync();
                }
            }

            var builder2 = CreateDistributedApplicationBuilder();
            var kafka2 = builder2.AddKafka("kafka");

            if (useVolume)
            {
                kafka2.WithDataVolume(volumeName);
            }
            else
            {
                kafka2.WithDataBindMount(bindMountPath!);
            }

            using (var app = builder2.Build())
            {
                await app.StartAsync();
                try
                {
                    var hb = Host.CreateApplicationBuilder();

                    hb.Configuration.AddInMemoryCollection(new Dictionary<string, string?>
                    {
                        [$"ConnectionStrings:{kafka2.Resource.Name}"] = await kafka2.Resource.ConnectionStringExpression.GetValueAsync(default)
                    });

                    hb.AddKafkaConsumer<string, string>("kafka", consumerBuilder =>
                    {
                        consumerBuilder.Config.GroupId = "aspire-consumer-group";
                        consumerBuilder.Config.AutoOffsetReset = AutoOffsetReset.Earliest;
                    });

                    using (var host = hb.Build())
                    {
                        await host.StartAsync();

                        var consumer = host.Services.GetRequiredService<IConsumer<string, string>>();
                        consumer.Subscribe(topic);
                        for (var i = 0; i < 10; i++)
                        {
                            var result = consumer.Consume(cts.Token);

                            Assert.Equal($"test-key", result.Message.Key);
                            Assert.Equal($"test-value{i}", result.Message.Value);
                        }
                    }
                }
                finally
                {
                    // Stops the container, or the Volume/mount would still be in use
                    await app.StopAsync();
                }
            }
        }
        finally
        {
            if (volumeName is not null)
            {
                DockerUtils.AttemptDeleteDockerVolume(volumeName);
            }

            if (bindMountPath is not null)
            {
                try
                {
                    File.Delete(bindMountPath);
                }
                catch
                {
                    // Don't fail test if we can't clean the temporary folder
                }
            }
        }
    }
```

### Exceptions (if any)

```
  Message: 
Confluent.Kafka.ConsumeException : Subscribed topic not available: test-topic: Broker: Unknown topic or partition

  Stack Trace: 
Consumer`2.Consume(Int32 millisecondsTimeout)
Consumer`2.Consume(CancellationToken cancellationToken)
KafkaFunctionalTests.WithDataShouldPersistStateBetweenUsages(Boolean useVolume) line 169
```

### .NET Version info

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Alirexaa**
## Description
Fixes #4909 
Alternate https://github.com/dotnet/aspire/pull/4947

## Checklist

- Is this feature complete?
  - [ ] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [ ] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [ ] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to aspire-docs issue: 
  - [ ] No

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5553)

### CODE REVIEW

**Comment by Alirexaa** [comment]
@eerhardt PTAL

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs [hunk 1]
```diff
@@ -101,12 +102,32 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+                    }
+                    else
+                    {
+                        // the docker container runs as a non-root user, so we need to grant other user's read/write permission
+                        // to the bind mount directory.
+                        const UnixFileMode BindMountPermissions =
+                            UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute |
+                            UnixFileMode.OtherRead | UnixFileMode.OtherWrite | UnixFileMode.OtherExecute;
+
+                        Directory.CreateDirectory(bindMountPath, BindMountPermissions);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:120]: The permissions have to happen after the create, due to `umask`. See 

https://github.com/dotnet/aspire/pull/5087/files#diff-faa1118f611b9b62c2f9a92ef4498a54aac418a5c8e569d9f5565b9d6353e473R90-R102

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs [hunk 2]
```diff
@@ -169,5 +170,7 @@ private static void ConfigureKafkaContainer(EnvironmentCallbackContext context,
             $"PLAINTEXT://{primaryEndpoint.Property(EndpointProperty.Host)}:29092,PLAINTEXT_HOST://{primaryEndpoint.Property(EndpointProperty.Host)}:{primaryEndpoint.Property(EndpointProperty.Port)},PLAINTEXT_INTERNAL://{internalEndpoint.Property(EndpointProperty.Host)}:{internalEndpoint.Property(EndpointProperty.Port)}");
 
         context.EnvironmentVariables["KAFKA_ADVERTISED_LISTENERS"] = advertisedListeners;
+
+        context.EnvironmentVariables["KAFKA_LOG_DIRS"] = Target;
```
Comment (eerhardt) [inline src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:174]: We only want to do this if `WithDataVolume` or `WithDataBindMount` was called, right?

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs [hunk 3]
```diff
@@ -139,6 +160,9 @@ await pipeline.ExecuteAsync(async token =>
                 {
                     // Stops the container, or the Volume/mount would still be in use
                     await app.StopAsync();
+                    //kafka shutdown has delay,so without delay to running instance using same data and second instance failed to start.
+                    await Task.Delay(TimeSpan.FromMinutes(1));
```
Comment (eerhardt) [inline tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:164]: Is there a way we can signal when this happens? Adding a time based delay is a recipe for flakey tests.

File: src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs (hunk 2 above)
Comment (Alirexaa) [reply src/Aspire.Hosting.Kafka/KafkaBuilderExtensions.cs:174]: I think so.

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs (hunk 3 above)
Comment (Alirexaa) [reply tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:164]: We have https://github.com/dotnet/aspire/issues/4878 for this but for the temporary solution, we could use `docker inspect` to check the container state but I couldn't find the container name in the app-model.

File: tests/Aspire.Hosting.Kafka.Tests/AddKafkaTests.cs [hunk 4]
```diff
@@ -109,6 +109,100 @@ public async Task VerifyManifest()
...
+        Assert.Equal(expectedManifest, manifest.ToString());
+    }
+
+    [Fact]
+    public async Task WithDataBindMountVerifyManifest()
+    {
+        using var appBuilder = TestDistributedApplicationBuilder.Create();
+
+        var kafka = appBuilder.AddKafka("kafka")
+            .WithDataBindMount("c:\\kafka-data");
```
Comment (davidfowl) [inline tests/Aspire.Hosting.Kafka.Tests/AddKafkaTests.cs:165]: What happens when this test runs on linux?

Comment (Alirexaa) [reply tests/Aspire.Hosting.Kafka.Tests/AddKafkaTests.cs]: I wrote a better test for this scenario.

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs (hunk 3 above)
Comment (davidfowl) [reply tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:153]: I think we might be able to improve this situation in these tests scenarios. I’ll take a look.

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs [hunk 5]
```diff
@@ -101,20 +97,34 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+                {
+                    // the docker container runs as a non-root user, so we need to grant other user's read/write permission
+                    // to the bind mount directory.
+                    // Note that we need to do this after creating the directory, because the umask is applied at the time of creation.
+                    const UnixFileMode BindMountPermissions =
+                        UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute |
+                        UnixFileMode.GroupRead | UnixFileMode.GroupWrite | UnixFileMode.GroupExecute |
+                        UnixFileMode.OtherRead | UnixFileMode.OtherWrite | UnixFileMode.OtherExecute;
+
+                    File.SetUnixFileMode(bindMountPath, BindMountPermissions);
```
Comment (davidfowl) [inline tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:113]: Is this something we need to do in the kafka extension and not in the tests? Pgadmin does something similar:

https://github.com/dotnet/aspire/blob/dad2369a4584dcf1b92d1ba957d424b18c373dbd/src/Aspire.Hosting.PostgreSQL/PostgresBuilderExtensions.cs#L194-L198

Comment (davidfowl) [reply tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:113]: I guess this is auto generated.

**Review by davidfowl — APPROVED** [review]

File: tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs (hunk 3 above)
Comment (Alirexaa) [reply tests/Aspire.Hosting.Kafka.Tests/KafkaFunctionalTests.cs:153]: Filed #5714

**Pull request merged**